pkg/operator/ceph/disruption/machinelabel/reconcile.go

```diff
@@ -0,0 +1,140 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Fetching the cephCluster
+	cephClusterInstance := &cephv1.CephCluster{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, cephClusterInstance)
+	if err != nil {
+		logger.Errorf("failed to fetch cephcluster %s: %+v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	// skipping the reconcile since the feature is switched off
+	if !cephClusterInstance.Spec.DisruptionManagement.ManageMachineDisruptionBudgets {
+		logger.Debugf("Skipping reconcile for cephCluster %s as manageMachineDisruption is turned off", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch list of machines available
+	machines := &mapiv1.MachineList{}
+	err = r.client.List(context.TODO(), machines, client.InNamespace(cephClusterInstance.Spec.DisruptionManagement.MachineDisruptionBudgetNamespace))
+	if err != nil {
+		logger.Errorf("failed tp fetch machine list %+v", machines)
+		return reconcile.Result{}, err
+	}
+
+	nodeMachineMap := map[string]machine{}
+
+	// Adding machines to nodeMachineMap
+	for _, m := range machines.Items {
+		if m.Status.NodeRef != nil {
+			nodeMachineMap[m.Status.NodeRef.Name] = machine{RawMachine: m}
+		}
+	}
+
+	// Marking machines that are occupied by the osd pods
+	for _, pod := range pods.Items {
+		if pod.Spec.NodeName != "" {
+			if machine, p := nodeMachineMap[pod.Spec.NodeName]; p {
+				machine.isOccupiedByOSD = true
+				nodeMachineMap[pod.Spec.NodeName] = machine
+			}
+		}
+	}
+
+	// Updating the machine status
+	for _, machine := range nodeMachineMap {
+		labels := machine.RawMachine.GetLabels()
+		clusterName, isClusterNamePresent := labels[MachineFencingLabelKey]
+		clusterNamespace, isClusterNamespacePresent := labels[MachineFencingNamespaceLabelKey]
+		if machine.isOccupiedByOSD {
+			if isClusterNamePresent && isClusterNamespacePresent && clusterName == request.Name && clusterNamespace == request.Namespace {
+				continue
+			}
+			labels[MachineFencingLabelKey] = request.Name
+			labels[MachineFencingNamespaceLabelKey] = request.Namespace
+			machine.RawMachine.SetLabels(labels)
+			err = r.client.Update(context.TODO(), &machine.RawMachine)
+			if err != nil {
+				logger.Errorf("failed to update machine %+v", err)
+				return reconcile.Result{}, err
+			}
+		} else {
+			if isClusterNamePresent && isClusterNamespacePresent && clusterName == "" && clusterNamespace == "" {
+				continue
+			}
+			labels[MachineFencingLabelKey] = ""
+			labels[MachineFencingNamespaceLabelKey] = ""
+			machine.RawMachine.SetLabels(labels)
+			err = r.client.Update(context.TODO(), &machine.RawMachine)
+			if err != nil {
+				logger.Errorf("failed to update machine %+v", err)
+				return reconcile.Result{}, err
+			}
+		}
+		logger.Infof("Successfully updated the Machine %s", machine.RawMachine.GetName())
```

Comment:
Similar to earlier question about frequency, this should be debug if it's frequent.

Reply:
it depends on the events of OSD pods.